format/macho/macho.go

```diff
@@ -0,0 +1,492 @@
+		d.FieldS32("cputype", cpuTypes)
+		d.FieldS32("cpusubtype")
+		// TODO ask about how to symmap this as it depends on a pair of values
+		d.FieldU32("filetype") // TODO expand this
+		ncmds = d.FieldU32("ncdms")
+		d.FieldU32("sizeofncdms")
+		d.FieldU32("flags") // TODO expand flags
+		if archBits == 64 {
+			d.FieldRawLen("reserved", 4*8, d.BitBufIsZero())
+		}
+	})
+	ncmdsIdx := 0
+	d.FieldStructArrayLoop("load_command", "load_command", func() bool {
+		return ncmdsIdx < int(ncmds)
+	}, func(d *decode.D) {
+		cmd := d.FieldU32("cmd", loadCommands)
+		cmdsize := d.FieldU32("cmdsize")
+		switch cmd {
+		case LC_UUID:
+			d.FieldStruct("uuid_command", func(d *decode.D) {
+				d.FieldRawLen("uuid", 16*8)
+			})
+		case LC_SEGMENT, LC_SEGMENT_64:
+			// nsect := (cmdsize - uint64(archBits)) / uint64(archBits)
+			var nsects uint64
+			if archBits == 32 {
+				d.FieldStruct("segment_command", func(d *decode.D) {
+					d.FieldRawLen("segname", 16*8)
+					d.FieldU32("vmaddr")
+					d.FieldU32("vmsize")
+					d.FieldU32("fileoff")
+					d.FieldU32("tfilesize")
+					d.FieldS32("initprot")
+					d.FieldS32("maxprot")
+					nsects = d.FieldU32("nsects")
+					d.FieldU32("flags") // TODO expand flags
+				})
+			} else {
+				d.FieldStruct("segment_command_64", func(d *decode.D) {
+					d.FieldStrFn("segname", func(d *decode.D) string {
+						return string(d.BytesLen(16))
+					})
+					d.FieldU64("vmaddr")
+					d.FieldU64("vmsize")
+					d.FieldU64("fileoff")
+					d.FieldU64("tfilesize")
+					d.FieldS32("initprot")
+					d.FieldS32("maxprot")
+					nsects = d.FieldU32("nsects")
+					d.FieldU32("flags") // TODO expand flags
+				})
+			}
+			var nsectIdx uint64
+			d.FieldStructArrayLoop("section", "section", func() bool {
+				return nsectIdx < nsects
+			},
+				func(d *decode.D) {
+					d.FieldStrFn("sectname", func(d *decode.D) string {
+						return string(d.BytesLen(16))
+					})
+					d.FieldStrFn("segname", func(d *decode.D) string {
+						return string(d.BytesLen(16))
+					})
+					if archBits == 32 {
+						d.FieldU32("address")
+						d.FieldU32("size")
+					} else {
+						d.FieldU64("address")
+						d.FieldU64("size")
+					}
+					d.FieldU32("offset")
+					d.FieldU32("align")
+					d.FieldU32("reloff")
+					d.FieldU32("nreloc")
+					d.FieldU32("flags") // TODO expand flags
+					d.FieldU32("reserved1")
+					d.FieldU32("reserved2")
+					if archBits == 64 {
+						d.FieldU32("reserved3")
+					}
+					nsectIdx++
+				})
+		case LC_TWOLEVEL_HINTS:
+			d.FieldU32("offset")
+			d.FieldU32("nhints")
+		case LC_LOAD_DYLIB, LC_ID_DYLIB, LC_LOAD_UPWARD_DYLIB, LC_LOAD_WEAK_DYLIB, LC_LAZY_LOAD_DYLIB, LC_REEXPORT_DYLIB:
+			d.FieldStruct("dylib_command", func(d *decode.D) {
+				offset := d.FieldU32("offset")
+				d.FieldU32("timestamp") // TODO human readable
+				d.FieldU32("current_version")
+				d.FieldU32("compatibility_version")
+				d.FieldStrFn("name", func(d *decode.D) string {
```

Comment:
Could this be `d.FieldUTF8("name", int(cmdsize) - int(offset))`?

Reply:
It think `d.FieldUTF8NullFixedLen("name", int(cmdsize)-int(offset))` is a better fit here. Taking that path.

Reply:
Good, as long as the string looks user friendly, no nulls or extra spaces etc